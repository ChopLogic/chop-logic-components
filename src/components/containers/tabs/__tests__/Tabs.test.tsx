import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { describe, expect, it, vi } from 'vitest';

import ChopLogicTabButton from '../elements/TabButton';
import ChopLogicTabContent from '../elements/TabContent';
import ChopLogicTabs, { ChopLogicTabsMode } from '../Tabs';

describe('ChopLogicTabs', () => {
  const tabs = [
    { id: 'tab-1', title: 'Tab 1', disabled: false, content: <>Tab 1 content</> },
    { id: 'tab-2', title: 'Tab 2', disabled: false, content: <>Tab 2 content</> },
    { id: 'tab-3', title: 'Tab 3', disabled: false, content: <>Tab 3 content</> },
  ];
  const testProps = {
    tabs,
    defaultTabId: 'tab-1',
    mode: 'horizontal' as ChopLogicTabsMode,
  };

  it('should match the snapshot', () => {
    const { asFragment } = render(<ChopLogicTabs {...testProps} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render another tab content when user clicks on tabs', async () => {
    render(<ChopLogicTabs {...testProps} />);
    expect(screen.getByText('Tab 1 content')).toBeInTheDocument();

    await userEvent.click(screen.getByText('Tab 2'));
    expect(screen.queryByText('Tab 1 content')).not.toBeInTheDocument();
    expect(screen.getByText('Tab 2 content')).toBeInTheDocument();

    await userEvent.click(screen.getByText('Tab 3'));
    expect(screen.queryByText('Tab 2 content')).not.toBeInTheDocument();
    expect(screen.getByText('Tab 3 content')).toBeInTheDocument();

    await userEvent.click(screen.getByText('Tab 1'));
    expect(screen.queryByText('Tab 3 content')).not.toBeInTheDocument();
    expect(screen.getByText('Tab 1 content')).toBeInTheDocument();
  });

  it('should render another tab content when user switches tabs via keyboard', async () => {
    render(<ChopLogicTabs {...testProps} />);
    expect(screen.getByText('Tab 1 content')).toBeInTheDocument();

    const firstTab = screen.getByText('Tab 1');
    firstTab.focus();
    await userEvent.keyboard('[ArrowRight]');
    expect(screen.queryByText('Tab 1 content')).not.toBeInTheDocument();
    expect(screen.getByText('Tab 2 content')).toBeInTheDocument();

    await userEvent.keyboard('[ArrowRight]');
    expect(screen.queryByText('Tab 2 content')).not.toBeInTheDocument();
    expect(screen.getByText('Tab 3 content')).toBeInTheDocument();

    await userEvent.keyboard('[ArrowRight]');
    expect(screen.queryByText('Tab 3 content')).not.toBeInTheDocument();
    expect(screen.getByText('Tab 1 content')).toBeInTheDocument();

    await userEvent.keyboard('[ArrowLeft]');
    expect(screen.queryByText('Tab 1 content')).not.toBeInTheDocument();
    expect(screen.getByText('Tab 3 content')).toBeInTheDocument();
  });
});

describe('ChopLogicTabButton', () => {
  const testProps = {
    title: 'Tab button',
    tabId: 'id-1',
    onTabSelect: vi.fn(),
    tabPanelId: 'panel-id-1',
    isSelected: true,
    mode: 'vertical' as ChopLogicTabsMode,
    isDisabled: false,
  };

  it('should match the snapshot', () => {
    const { asFragment } = render(<ChopLogicTabButton {...testProps} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should have the aria attribute', () => {
    render(<ChopLogicTabButton {...testProps} />);
    expect(screen.getByRole('tab')).toBeInTheDocument();
  });
});

describe('ChopLogicTabContent', () => {
  const tabs = [
    { id: 'tab-1', title: 'Tab 1', disabled: false, content: <>Tab 1 content</> },
    { id: 'tab-2', title: 'Tab 2', disabled: false, content: <>Tab 2 content</> },
    { id: 'tab-3', title: 'Tab 3', disabled: false, content: <>Tab 3 content</> },
  ];

  it('should match the snapshot for the first tab', () => {
    const { asFragment } = render(<ChopLogicTabContent tabs={tabs} selectedTabId='tab-1' />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should match the snapshot for the second tab', () => {
    const { asFragment } = render(<ChopLogicTabContent tabs={tabs} selectedTabId='tab-2' />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should have the aria attribute', () => {
    render(<ChopLogicTabContent tabs={tabs} selectedTabId='tab-2' />);
    expect(screen.getByRole('tabpanel')).toBeInTheDocument();
  });
});
