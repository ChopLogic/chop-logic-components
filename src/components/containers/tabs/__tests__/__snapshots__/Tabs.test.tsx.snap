// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ChopLogicTabButton > should match the snapshot 1`] = `
<DocumentFragment>
  <button
    aria-controls="panel-id-1"
    aria-selected="true"
    class="sc-blHHSb gLYyzF"
    id="id-1"
    role="tab"
  >
    Tab button
  </button>
</DocumentFragment>
`;

exports[`ChopLogicTabContent > should match the snapshot for the first tab 1`] = `
<DocumentFragment>
  <div
    aria-labelledby="tab-1"
    class="sc-gtLWhw edHYGl"
    id="tabpanel_tab-1"
    role="tabpanel"
  >
    Tab 1 content
  </div>
</DocumentFragment>
`;

exports[`ChopLogicTabContent > should match the snapshot for the second tab 1`] = `
<DocumentFragment>
  <div
    aria-labelledby="tab-2"
    class="sc-gtLWhw edHYGl"
    id="tabpanel_tab-2"
    role="tabpanel"
  >
    Tab 2 content
  </div>
</DocumentFragment>
`;

exports[`ChopLogicTabs > should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-fAUdSK juRevX test-class"
    id="tabs-id"
  >
    <div
      class="sc-egkSDF iLuVKo"
      role="tablist"
    >
      <button
        aria-controls="tabpanel_tab-1"
        aria-selected="true"
        class="sc-blHHSb jyZlgZ"
        id="tab-1"
        role="tab"
      >
        Tab 1
      </button>
      <button
        aria-controls="tabpanel_tab-2"
        aria-selected="false"
        class="sc-blHHSb lctyDo"
        id="tab-2"
        role="tab"
      >
        Tab 2
      </button>
      <button
        aria-controls="tabpanel_tab-3"
        aria-selected="false"
        class="sc-blHHSb lctyDo"
        id="tab-3"
        role="tab"
      >
        Tab 3
      </button>
    </div>
    <div
      aria-labelledby="tab-1"
      class="sc-gtLWhw edHYGl"
      id="tabpanel_tab-1"
      role="tabpanel"
    >
      Tab 1 content
    </div>
  </div>
</DocumentFragment>
`;
